pubsub.go

```diff
@@ -194,8 +194,11 @@ type PubSubRouter interface {
 	// Attach is invoked by the PubSub constructor to attach the router to a
 	// freshly initialized PubSub instance.
 	Attach(*PubSub)
-	// AddPeer notifies the router that a new peer has been connected.
-	AddPeer(peer.ID, protocol.ID)
+	// AddPeer notifies the router that a new peer has been connected. It
+	// includes a reference to the initial RPC that will be sent to the peer.
+	// Routers may add messages to the RPC to try to have them sent in the
+	// initial packet. This is also referred to as the "hello packet."
+	AddPeer(peer.ID, protocol.ID, *RPC)
```

Comment:
Making a param an output is very weird. What I usually expect from a function is that the params are inputs and the returned value is an output.

What you can do with this function is that you can return the `*pb.ControlExtensions` instead which I think is cleaner and you don't have to worry about mutating the parameter.

```suggestion
	AddPeer(peer.ID, protocol.ID) *pb.ControlExtensions
```

Reply:
I tried this out, and I think it's worse than the out param. It leads to all pubsub implementations having to know about control extensions in their signatures. Also there's no particular reason to only allow the ControlExtensions besides it's the only thing we bundle in the first message. Passing the out param is more general.

Instead I'll follow the append style where I return a (possibly new) pointer.